### ISSUE TITLE
sql: ALTER COLUMN .. . SET DATA TYPE doesn't check DEFAULT clause

### ISSUE THREAD

**Issue by rafiss** [cockroachdb/cockroach#70605]
**Describe the problem**

It is possible to alter a column's type in a way that is incompatible with the `DEFAULT` clause.

**To Reproduce**

```
create table t (a int primary key, b timestamp default current_timestamp());
alter table t alter column b set data type timestamptz;
insert into t values (1);
```

This results in: `ERROR: argument of  must be type timestamptz, not type timestamp`.

```
> show create table t;
  table_name |                        create_statement
-------------+------------------------------------------------------------------
  t          | CREATE TABLE public.t (
             |     a INT8 NOT NULL,
             |     b TIMESTAMPTZ NULL DEFAULT current_timestamp():::TIMESTAMP,
             |     CONSTRAINT "primary" PRIMARY KEY (a ASC),
             |     FAMILY "primary" (a, b)
             | )
```

**Expected behavior**
Either the `ALTER` should be disallowed, or we should implement casting logic so the example above can be handled without error.

**Environment:**
I was using `master` but this affects v21.1+

Epic CRDB-8948

**Comment by RichardJCai** [comment]
Postgres disallows the cast based on the default expression.
```
rcai=# create table testalter2(x text default 'hello', y int);
CREATE TABLE
rcai=# alter table testalter2 alter column x type int;
ERROR:  column "x" cannot be cast automatically to type integer
HINT:  You might need to specify "USING x::integer".
rcai=# alter table testalter2 alter column x type int USING x::integer;
ERROR:  default for column "x" cannot be cast automatically to type integer
rcai=# alter table testalter2 alter column x type int;
```

**Comment by otan** [comment]
need to sneak something like

```


	if col.HasDefault() {
		expr, err := parser.ParseExpr(col.GetDefaultExpr())
		if err != nil {
			return err
		}
		if _, _, err := sanitizeColumnExpression(params, expr, typ, "ALTER COLUMN TYPE"); err != nil {
			if t.Using == nil {
				return errors.WithHintf(
					errors.Wrapf(err, "cannot change column type with incompatible DEFAULT expression"),
					`you may need to specify "USING %s::%s"`,
					col.ColName().String(),
					typ.SQLString(),
				)
			}

		}
	}
```

**Comment by ajwerner** [comment]
@rafiss is it fair to say that experience is taking this? If not, let us know and put it back on our board.

**Comment by mgartner** [comment]
I looked into how assignment casts and DEFAULT expressions should behave, and it is more or less "intuitive" if you have an understanding of assignment casts. I say "intuitive" sardonically because assignment casts are really quite complicated. Before reading the rest of this comment, I'd recommend reading [this PR description](https://github.com/cockroachdb/cockroach/pull/70722#issue-1006808857) and [these code comments](https://github.com/cockroachdb/cockroach/blob/128ee06f44f304fd67a7fc277fe7e069f4e49257/pkg/sql/sem/tree/cast.go#L44-L136) for a primer on assignment casts.

### TLDR

**To summarize the behavior of assignment casts, DEFAULT expressions, and `ALTER TABLE .. ALTER COLUMN .. TYPE` statement:**

1. A column's default expression must have a type that can be assignment cast to the column's type.
2. On insert, assigment casts are performed on DEFAULT expressions if their type is not identical to the column's type.
3. `ALTER TABLE .. ALTER COLUMN .. TYPE ..` fails if there is not a valid assignment cast from the column's existing type to the target type.
3. During `ALTER TABLE .. ALTER COLUMN .. TYPE ..`, existing values undergo an assignment cast to be converted to the new target type. These assignment casts can fail, in which case the `ALTER TABLE` statement fails.

### Examples that prove this behavior

DEFAULT expressions undergo assignment casts during `INSERT`s if their type does not match their column's type, similarly to how insert values undergo assignment casts to match column types.

In this example, you can see how the values for `d` are rounded to match the column type:

```
marcus=# CREATE TABLE t (k INT PRIMARY KEY, d DECIMAL(10, 0) DEFAULT 123.45::DECIMAL(10, 2));
CREATE TABLE

marcus=# INSERT INTO t VALUES (1);
INSERT 0 1

marcus=# INSERT INTO t VALUES (2, 987.65);
INSERT 0 1

marcus=# SELECT * FROM t;
 k |  d
---+-----
 1 | 123
 2 | 988
(2 rows)
```

We know that an assignment cast, not an explicit cast, is applied to this DEFAULT expression because it does not perform width truncation:

```
marcus=# CREATE TABLE t (k INT PRIMARY KEY, c CHAR DEFAULT 'foo'::CHAR(3));
CREATE TABLE

marcus=# INSERT INTO t VALUES (1);
ERROR:  22001: value too long for type character(1)
LOCATION:  bpchar, varchar.c:311

marcus=# INSERT INTO t VALUES (1, 'foo'::char(3));
ERROR:  22001: value too long for type character(1)
LOCATION:  bpchar, varchar.c:311
```

Notice that an explicit cast is able to truncate widths without issue:

```
marcus=# SELECT 'foo'::CHAR(3)::CHAR;
 bpchar
--------
 f
(1 row)
```

When a column type is altered, the existing values of the column undergo an assignment cast to be converted to the new type. You can tell this is the case because the assignment cast of `'foo'::CHAR(3)` to `CHAR` fails because it's width cannot be truncated. Once we delete this value from the table, the `ALTER TABLE` command succeeds.

```
marcus=# CREATE TABLE t (k INT PRIMARY KEY, c CHAR(3) DEFAULT 'foo'::CHAR(3));
CREATE TABLE

marcus=# INSERT INTO t VALUES (1);
INSERT 0 1

marcus=# SELECT * FROM t;
 k |  c
---+-----
 1 | foo
(1 row)

marcus=# ALTER TABLE t ALTER COLUMN c TYPE CHAR;
ERROR:  22001: value too long for type character(1)
LOCATION:  bpchar, varchar.c:311

marcus=# DELETE FROM t;
DELETE 1

marcus=# ALTER TABLE t ALTER COLUMN c TYPE CHAR;
ALTER TABLE
```

The default expression is not altered, so performing an insert into this table without specifying a value for `c` gives us an error:

```
marcus=# \d t
                          Table "public.t"
 Column |     Type     | Collation | Nullable |       Default
--------+--------------+-----------+----------+---------------------
 k      | integer      |           | not null |
 c      | character(1) |           |          | 'foo'::character(3)
Indexes:
    "t_pkey" PRIMARY KEY, btree (k)

marcus=# INSERT INTO t VALUES (1);
ERROR:  22001: value too long for type character(1)
LOCATION:  bpchar, varchar.c:311
```

If an assignment cast from the existing column type to the target column type does not exist, the `ALTER TABLE` will fail. The example below shows this because there is no assignment cast from `TEXT` to `INT4`.

```
marcus=# CREATE TABLE t (t TEXT);
CREATE TABLE

marcus=# ALTER TABLE t ALTER COLUMN t TYPE INT4;
ERROR:  42804: column "t" cannot be cast automatically to type integer
HINT:  You might need to specify "USING t::integer".
LOCATION:  ATPrepAlterColumnType, tablecmds.c:11202
```

@RichardJCai stumbled upon an interesting case above that is worth highlighting with an example that does not use `USING`, to make things a bit simpler. The `ALTER TABLE` below fails, but not because there isn't an assignment cast from the exiting column type `TIME` to the target type `INTERVAL` - this is a valid assignment cast. However, Postgres also checks that a valid assignment cast exists from the default expression type `TIMETZ` to the target type `INTERVAL`. Such an assignment cast does not exist, so the `ALTER TABLE` fails immediately.

```
marcus=# CREATE TABLE t (k INT PRIMARY KEY, t TIME DEFAULT now()::TIMETZ);
CREATE TABLE

marcus=# ALTER TABLE t ALTER COLUMN t TYPE INTERVAL;
ERROR:  42804: default for column "t" cannot be cast automatically to type interval
LOCATION:  ATExecAlterColumnType, tablecmds.c:11507
```

Similarly, you cannot create a column if there is no valid assignment cast from its DEFAULT expression's type to its type:

```
marcus=# CREATE TABLE t (i INTERVAL DEFAULT now()::TIMETZ);
ERROR:  42804: column "i" is of type interval but default expression is of type time with time zone
HINT:  You will need to rewrite or cast the expression.
LOCATION:  cookDefault, heap.c:3071
```

**Comment by mgartner** [comment]
> 2. On insert, assignment casts are performed on DEFAULT expressions if their type is not identical to the column's type.

FYI, I'll take care of this while I continue to implement assignment casts.

**Comment by rafiss** [comment]
thanks @mgartner! your write up is very helpful

interestingly, it looks like CRDB just goes ahead and automatically converts the existing values if it thinks it can (unlike postgres, which would return an error at the ALTER statement)
```
root@:26257/defaultdb> CREATE TABLE t (k INT PRIMARY KEY, c CHAR(3) DEFAULT 'foo'::CHAR(3));
CREATE TABLE

root@:26257/defaultdb> INSERT INTO t VALUES (1);
INSERT 1

root@:26257/defaultdb> set enable_experimental_alter_column_type_general='true';
SET

root@:26257/defaultdb> ALTER TABLE t ALTER COLUMN c TYPE CHAR;
NOTICE: ALTER COLUMN TYPE changes are finalized asynchronously; further schema changes on this table may be restricted until the job completes; some writes to the altered column may be rejected until the schema change is finalized
ALTER TABLE

root@:26257/defaultdb> select * from t;
  k | c
----+----
  1 | f
```

anyway for now, before  you get to the assignment cast support, i think we should focus on improving the error handling (in https://github.com/cockroachdb/cockroach/pull/71423)
